File: apps/agent/src/ingest/chunker.ts
const TARGET_CHARS = 720;
const HARD_MAX = 1100;
const OVERLAP_CHARS = 140;

export interface Chunk {
  text: string;
  ordinal: number;
}

export function chunk(text: string): Chunk[] {
  const normalized = text.replace(/\r\n/g, "\n").trim();
  if (!normalized) return [];

  const paragraphs = normalized
    .split(/\n{2,}/)
    .map((p) => p.trim())
    .filter(Boolean);

  const chunks: Chunk[] = [];
  let buffer = "";
  let ordinal = 0;

  const flush = () => {
    const t = buffer.trim();
    if (!t) return;
    chunks.push({ text: t, ordinal: ordinal++ });
    buffer = OVERLAP_CHARS > 0 ? tailOverlap(t, OVERLAP_CHARS) : "";
  };

  for (const para of paragraphs) {
    if (para.length > HARD_MAX) {
      if (buffer) flush();
      for (const piece of splitSentence(para, TARGET_CHARS, HARD_MAX)) {
        buffer = (buffer ? buffer + "\n\n" : "") + piece;
        if (buffer.length >= TARGET_CHARS) flush();
      }
      continue;
    }
    const next = buffer ? buffer + "\n\n" + para : para;
    if (next.length > HARD_MAX) {
      flush();
      buffer = buffer ? buffer + para : para;
    } else {
      buffer = next;
    }
    if (buffer.length >= TARGET_CHARS) flush();
  }
  if (buffer.trim()) flush();
  return chunks;
}

function tailOverlap(text: string, n: number): string {
  if (text.length <= n) return text;
  const slice = text.slice(text.length - n);
  const breakAt = slice.search(/[.!?]\s|\n/);
  return breakAt >= 0 ? slice.slice(breakAt + 1).trim() : slice.trim();
}

function splitSentence(para: string, target: number, hardMax: number): string[] {
  const sentences = para.split(/(?<=[.!?])\s+(?=[A-Z0-9"'])/g);
  const out: string[] = [];
  let buf = "";
  for (const s of sentences) {
    const candidate = buf ? buf + " " + s : s;
    if (candidate.length > hardMax) {
      if (buf) out.push(buf);
      if (s.length > hardMax) {
        for (let i = 0; i < s.length; i += target) {
          out.push(s.slice(i, i + target));
        }
        buf = "";
      } else {
        buf = s;
      }
    } else {
      buf = candidate;
    }
    if (buf.length >= target) {
      out.push(buf);
      buf = "";
    }
  }
  if (buf) out.push(buf);
  return out;
}

File: apps/agent-py/app/ingest/chunker.py
"""Paragraph-aware chunker — port of apps/agent/src/ingest/chunker.ts."""
from __future__ import annotations

import re

TARGET_CHARS = 720
HARD_MAX = 1100
OVERLAP_CHARS = 140


def _tail_overlap(text: str, n: int) -> str:
    if len(text) <= n:
        return text
    sl = text[len(text) - n:]
    m = re.search(r"[.!?]\s|\n", sl)
    return sl[m.start() + 1:].strip() if m else sl.strip()


def _split_sentence(para: str, target: int, hard_max: int) -> list[str]:
    sentences = re.split(r"(?<=[.!?])\s+(?=[A-Z0-9\"'])", para)
    out: list[str] = []
    buf = ""
    for s in sentences:
        candidate = (buf + " " + s) if buf else s
        if len(candidate) > hard_max:
            if buf:
                out.append(buf)
            if len(s) > hard_max:
                for i in range(0, len(s), target):
                    out.append(s[i:i + target])
                buf = ""
            else:
                buf = s
        else:
            buf = candidate
        if len(buf) >= target:
            out.append(buf)
            buf = ""
    if buf:
        out.append(buf)
    return out


def chunk(text: str) -> list[dict]:
    normalized = text.replace("\r\n", "\n").strip()
    if not normalized:
        return []
    paragraphs = [p.strip() for p in re.split(r"\n{2,}", normalized) if p.strip()]

    chunks: list[dict] = []
    state = {"buffer": "", "ordinal": 0}

    def flush():
        t = state["buffer"].strip()
        if not t:
            return
        chunks.append({"text": t, "ordinal": state["ordinal"]})
        state["ordinal"] += 1
        state["buffer"] = _tail_overlap(t, OVERLAP_CHARS) if OVERLAP_CHARS > 0 else ""

    for para in paragraphs:
        if len(para) > HARD_MAX:
            if state["buffer"]:
                flush()
            for piece in _split_sentence(para, TARGET_CHARS, HARD_MAX):
                state["buffer"] = (state["buffer"] + "\n\n" + piece) if state["buffer"] else piece
                if len(state["buffer"]) >= TARGET_CHARS:
                    flush()
            continue
        nxt = (state["buffer"] + "\n\n" + para) if state["buffer"] else para
        if len(nxt) > HARD_MAX:
            flush()
            state["buffer"] = (state["buffer"] + para) if state["buffer"] else para
        else:
            state["buffer"] = nxt
        if len(state["buffer"]) >= TARGET_CHARS:
            flush()
    if state["buffer"].strip():
        flush()
    return chunks
